lib/view_model/active_record/update_operation.rb

```diff
@@ -526,35 +540,121 @@ def build_updates_for_has_many_through(association_data, reference_strings, upda
+                "Stale functional update for association '#{association_data.direct_reflection.name}' - "\
+                "could not match referenced viewmodels: [#{ref_by_indirect_vm_ref.keys.map(&:to_s).join(', ')}]")
+            end
+
+          else
+            raise ArgumentError.new("Unknown functional update: '#{fupdate.class}'")
+          end
+        end
+
+      else
+        raise_deserialization_error("Unknown association_update type '#{association_update.class.name}'")
+      end
+
+      new_through_viewmodels = target_indirect_viewmodels.map do |target|
+        target_viewmodel_ref, ref  = target.values_at(:viewmodel_ref, :ref)
 
         existing_through_viewmodel =
-          previous_through_viewmodels_by_indirect_ref[target_reference].try(:shift) if target_reference
+          (previous_through_viewmodels_by_indirect_ref[target_viewmodel_ref]&.shift if target_viewmodel_ref)
+
+        unless ref || existing_through_viewmodel
+          raise_deserialization_error('Existing through model was not resolved')
+        end
 
-        existing_through_viewmodel || through_viewmodel_class.for_new_model
+        {
+          viewmodel: existing_through_viewmodel || through_viewmodel_class.for_new_model,
+          ref:       ref,
+        }
       end
 
       orphaned_through_viewmodels = previous_through_viewmodels_by_indirect_ref.flat_map { |_, vms| vms }
 
-      if new_through_viewmodels != previous_through_viewmodels
+      if new_through_viewmodels.map { |vm| vm[:viewmodel] } != previous_through_viewmodels
         self.association_changed!(association_data.association_name)
       end
 
       positions = Array.new(new_through_viewmodels.length)
       if through_viewmodel_class._list_member?
         # It's always fine to use a position, since this is always owned data (no moves, so no positions to ignore.)
-        get_position = ->(index)     { new_through_viewmodels[index]._list_attribute }
+        get_position = ->(index)     { new_through_viewmodels[index][:viewmodel]._list_attribute }
         set_position = ->(index, pos){ positions[index] = pos }
 
         ActsAsManualList.update_positions((0...new_through_viewmodels.size).to_a, # indexes
                                           position_getter: get_position,
                                           position_setter: set_position)
       end
 
-      new_through_updates = new_through_viewmodels.zip(reference_strings, positions).map do |viewmodel, ref, position|
+      new_through_updates = new_through_viewmodels.zip(positions).map do |new_viewmodel, position|
+        viewmodel, ref = new_viewmodel.values_at(:viewmodel, :ref)
+
         update_data = UpdateData.empty_update_for(viewmodel)
-        update_data.referenced_associations[indirect_reflection.name] = ref # TODO layering violation
+        update_data.referenced_associations[indirect_reflection.name] = ref if ref # TODO layering violation
```

Comment:
When do we not have a `ref` here? If it's a previous child that hasn't been updated?